cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java

```diff
@@ -122,6 +147,101 @@ public void createDataset(String datasetInstanceName,
   }
 
   @Override
+  public <T> T usePlugin(String pluginType, String pluginName, PluginProperties properties) {
+    return usePlugin(pluginType, pluginName, properties, new PluginSelector());
+  }
+
+  @Override
+  public <T> T usePlugin(final String pluginType, final String pluginName,
+                         PluginProperties properties, final PluginSelector selector) {
+
+    Preconditions.checkArgument(hasPlugin(pluginType, pluginName),
+                                "Plugin of type {}, name {} was already added.", pluginType, pluginName);
+    Preconditions.checkArgument(properties != null, "Plugin properties cannot be null");
+
+    Map.Entry<PluginInfo, PluginClass> pluginEntry = pluginRepository.findPlugin(adapterConfig.getTemplate(),
+                                                                                 pluginType, pluginName, selector);
+    if (pluginEntry == null) {
+      return null;
+    }
+    try {
+      T instance = pluginInstantiator.newInstance(pluginEntry.getKey(), pluginEntry.getValue(), properties);
+      registerPlugin(pluginEntry.getKey(), pluginEntry.getValue(), properties);
+      return instance;
+    } catch (IOException e) {
+      // If the plugin jar is deleted without notifying the adapter service.
+      return null;
+    } catch (ClassNotFoundException e) {
+      // Shouldn't happen
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public <T> Class<T> usePluginClass(String pluginType, String pluginName, PluginProperties properties) {
+    return usePluginClass(pluginType, pluginName, properties, new PluginSelector());
+  }
+
+  @Override
+  public <T> Class<T> usePluginClass(String pluginType, String pluginName,
+                                     PluginProperties properties, PluginSelector selector) {
+
+    Preconditions.checkArgument(hasPlugin(pluginType, pluginName),
+                                "Plugin of type %s, name %s was already added.", pluginType, pluginName);
```

Comment:
If `hasPlugin` return true, meaning the plugin is already added in this configurer from some prior call to one of those `usePlugin*` method. This is to avoid adding the same plugin with different properties (and potentially have the selector to select a different plugin).